### ISSUE TITLE
Download reference failed

### ISSUE THREAD

**Issue by lh12565** [LTLA/SingleR#72]
Hi,

 I want to get the "DatabaseImmuneCellExpressionData" dataset using "DatabaseImmuneCellExpressionData()" function, but I couldn't download it and report following error:

```
Warning messages:
1: download failed
  web resource path: ‘https://experimenthub.bioconductor.org/fetch/3112’
  local file path: ‘/data/.cache/ExperimentHub/dbd05d1658b9_3112’
  reason: Timeout was reached: [s3.amazonaws.com] Connection timed out after 10002 milliseconds
2: bfcadd() failed; resource removed
  rid: BFC45
  fpath: ‘https://experimenthub.bioconductor.org/fetch/3112’
  reason: download failed
3: download failed
  hub path: ‘https://experimenthub.bioconductor.org/fetch/3112’
  cache resource: ‘EH3096 : 3112’
  reason: bfcadd() failed; see warnings()
```

 I wonder if you would consider upload a copy (RData or RDS) for me.
Thanks!

**Comment by dtm2451** [comment]
Hi @lh12565,

Your issue looks potentially like a connection issue somewhere along the path between your system and where this data is remotely stored.  Something that might have been a temporary thing.

This makes me want to suggest trying again.  Let us know if you get the same error when you do so.  Running `DatabaseImmuneCellExpressionData()` works for me today.

**Comment by lh12565** [comment]
Hi

> Hi @lh12565,
> 
> Your issue looks potentially like a connection issue somewhere along the path between your system and where this data is remotely stored. Something that might have been a temporary thing.
> 
> This makes me want to suggest trying again. Let us know if you get the same error when you do so. Running `DatabaseImmuneCellExpressionData()` works for me today.

I tried many times and still reported the same error.

**Comment by LTLA** [comment]
I'm willing to bet it's a firewall issue. What happens if you navigate to:

https://experimenthub.bioconductor.org/fetch/3112

via your browser? Something starts downloading for me.

If you do have a firewall and you know your proxy server (your IT support team can probably help here), you can set the `PROXY` option via `setExperimentHubOption` and see if it helps. 

As I mentioned off-line, I don't want to send you files, because then everyone with a firewall will ask me to send them files, and I don't want to be spending time sending files, especially if the files update and I have to send across a bunch of new files. Much better to solve this properly.

**Comment by saisaitian** [comment]
Dear author，could  you tell me how to  set the PROXY option via setExperimentHubOption, I have no idea

**Comment by LTLA** [comment]
Read `?setExperimentHubOption`.

**Comment by saisaitian** [comment]
“PROXY”:
request object returned by httr::use_proxy(). The request object describes a proxy connection allowing Internet access, usually through a restrictive firewall. Setting this option sends all ExperimentHub requests through the proxy. Default: NULL.

In setHubOption("PROXY", value), value can be one of NULL, a request object returned by httr::use_proxy(), or a well-formed URL as character(1). The URL can be completely specified by http://username:password@proxy.dom.com:8080; username:password and port (e.g. :8080) are optional. 
In this datasets, url
should be?,please give us a example, thank you

**Comment by LTLA** [comment]
Well, I don't know your proxy server, so I don't have any meaningful example to give.

Can you access https://experimenthub.bioconductor.org/fetch/3112? Does anything happen?

**Comment by saisaitian** [comment]
I can download it ,named logcounts rds

**Comment by saisaitian** [comment]
However， when  I set  PROXY, there also is error
> setExperimentHubOption("PROXY", 'https://192.168.0.1')
> mouse_1<-ImmGenData()
Cannot determine internet connection.
 If you experience connection issues consider using 'localHub=TRUE'
snapshotDate(): 2019-10-22
see ?SingleR and browseVignettes('SingleR') for documentation
loading from cache
Error: failed to load resource
  name: EH3092
  title: ImmGen logcounts
  reason: error reading from connection

**Comment by LTLA** [comment]
That looks more like your modem gateway than a proxy server.

Anyway, do the following and report back which steps fail.

```r
library(ExperimentHub)
ehub <- ExperimentHub()

ehub[["EH3092"]] # Does this fail? If so, proceed.

library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)

# Does this fail? If so, proceed.
bfcadd(bfc, "EH3092 : 3108", fpath="https://experimenthub.bioconductor.org/fetch/3108")

library(httr)

# Does this fail?
GET("https://experimenthub.bioconductor.org/fetch/3108", 
    progress(con = stderr()), 
    write_disk("thing", overwrite = TRUE))
```

IIRC, ExperimentHub is based on AWS S3; this seems to have some issues in China (both in terms of speed and connectivity), which may - or may not - be relevant here.

**Comment by saisaitian** [comment]
> tmp=bfcadd(bfc, "EH3092 : 3108", fpath="https://experimenthub.bioconductor.org/fetch/3108")
  |==============================================================================| 100%，
but how can I import tmp?

 > tmp
                                                                                             BFC7 
"C:\\Users\\Administrator\\AppData\\Local\\BiocFileCache\\BiocFileCache\\Cache/542427474672_3108"

**Comment by LTLA** [comment]
Let's be absolutely clear.

This fails for you:

```r
library(ExperimentHub)
ehub <- ExperimentHub()
ehub[["EH3092"]]
```

This succeeds for you:

```r
library(BiocFileCache)
bfc <- BiocFileCache(ask=FALSE)
bfcadd(bfc, "EH3092 : 3108", fpath="https://experimenthub.bioconductor.org/fetch/3108")
```

**Please post the FULL error message from the first chunk of code.** Also post your session information.

**Comment by saisaitian** [comment]
First, this fails for me

```
> library(SingleR)
> ig_mm.se=ImmGenData()
snapshotDate(): 2019-10-22
see ?SingleR and browseVignettes('SingleR') for documentation
loading from cache
Error: failed to load resource
  name: EH3092
  title: ImmGen logcounts
  reason: error reading from connection
```
Then I follow your guide.

```
library(ExperimentHub)
ehub <- ExperimentHub()

tmp2=ehub[["EH3092"]]
```

**Comment by LTLA** [comment]
Interesting. This seems to be different from the error you originally described in #74.

I would ask you to try:

```r
AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, TRUE, TRUE)
```

and report the errors (if any) that occur.

Note to self (and maybe also @lshep): the logic is that the error message is coming from:

https://github.com/Bioconductor/AnnotationHub/blob/9eb4e7d50f00336ea19f3fc7b08063c20c2394d2/R/Hub-utils.R#L298-L307

and so we try to see where the error is *really* coming from by calling internals to evade the `try` block.

**Comment by lshep** [comment]
Thanks for keeping me posted.  Let me know if you come up with more and I will do the same - this seems like a tricky one but if we can figure it out will hopefully smooth things out greatly

**Comment by saisaitian** [comment]
It also failed for me. Could you provied me a link to download ImmGenData(), thank you

**Comment by LTLA** [comment]
> Could you provied me a link to download ImmGenData()

That's exactly the problem right now. The link is already there, in **ExperimentHub**, and if that's not working for you, then you will not be able to access any Bioconductor *Hub resources.

If you want to get it working, it is much better to stay patient and help me debug this, rather than asking for links that I have already said I will not provide. See my reasoning above.

If you run the `AnnotationHub:::.cache_internal` command and get the error, what does running `traceback()` immediately afterwards say?

**Comment by saisaitian** [comment]
Thank you, it always get errors like these below:
```
> ttt=AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, TRUE, TRUE)
snapshotDate(): 2019-10-22
downloading 1 resources
retrieving 1 resource
  |=====                                                                                     |   5%Error: 1 resources failed to download
In addition: Warning messages:
1: download failed
  web resource path: ‘https://experimenthub.bioconductor.org/fetch/3108’
  local file path: ‘C:\Users\Administrator\AppData\Local\ExperimentHub\ExperimentHub\Cache\file3a54f5f484e’
  reason: Recv failure: Connection was reset 
2: bfcdownload() failed
  rid: BFC4
  file: ‘https://experimenthub.bioconductor.org/fetch/3108’
  reason: download failed 
3: download failed
  hub path: ‘https://experimenthub.bioconductor.org/fetch/3108’
  cache resource: ‘EH3092 : 3108’
  reason: download failed; see warnings() 
> 
```

**Comment by LTLA** [comment]
There seems to be two problems here.

## First problem

This shows the "connection was reset" message described [above](https://github.com/LTLA/SingleR/issues/72#issuecomment-570035356). This is probably a network issue; probably on your side, given that I cannot reproduce it; and probably some kind of institutional firewall. Indeed, you can see that it was actually downloading stuff before the connection was lost, so some other process is intervening to stop the download from working.

To confirm this: try running the two commands below multiple times. (Just copy and paste them into your R session and run them both about 10 times). Is it only the `AnnotationHub:::.cache_internal` that fails? Or do both commands fail?

```r
AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, TRUE, TRUE)

GET("https://experimenthub.bioconductor.org/fetch/3108", 
     progress(con = stderr()), 
     write_disk("thing", overwrite = TRUE))
```

## Second problem

This shows the "error reading from connection" message described [above](https://github.com/LTLA/SingleR/issues/72#issuecomment-569837239). This seems to be something more specific to **AnnotationHub** given that the corresponding `bfcrpath()` command worked. 

To provide some more information, try running:

```r
# NOTE! This is different from the previous command!
AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, FALSE, TRUE)
traceback()
```

Hopefully this should trigger the second error.

**Comment by saisaitian** [comment]
Hi, thank you. Today , it worked. Althougth , I do not know why it worked, it really worked today.
```
> t1=AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, TRUE, TRUE)
snapshotDate(): 2019-10-22
downloading 1 resources
retrieving 1 resource
  |====================================================================================| 100%

loading from cache

> t2=GET("https://experimenthub.bioconductor.org/fetch/3108", 
+     progress(con = stderr()), 
+     write_disk("thing", overwrite = TRUE))
  |====================================================================================| 100%

```
Step2:
```
> AnnotationHub:::.cache_internal(ExperimentHub()["EH3092"], "", 100, FALSE, TRUE)
snapshotDate(): 2019-10-22
loading from cache
                                                                                           EH3092 
"C:\\Users\\Administrator\\AppData\\Local\\ExperimentHub\\ExperimentHub\\Cache/3b8c4b9377ef_3108" 
```

**Comment by LTLA** [comment]
So, certainly a network issue. Can you indicate your locale? Countries physically distant from AWS US-east region may have higher latency, and China seems to have more problems on top of that.

@lshep It may be worthwhile wrapping the `bfcneedsupdate()` call inside a `tryCatch` that just proceeds if the remote cannot be accessed; this should allow people to keep going if they were lucky enough to get a cached object at any point. Currently, even cached results require connectivity to use them, which seems to defeat the purpose.

**Comment by lshep** [comment]
I can look into this. For the hubs,  without internet connectivity, to use resources already downloaded and skip checks, you should be able to use localhub=true in the hub constructor.

**Comment by LTLA** [comment]
Sure, but seems like it would be a pretty easy usability win to switch to that mode automatically. And it would save me having to pass a `localHub` argument all the way down to the EHub constructor.

**Comment by lshep** [comment]
I'll look at it this week.

**Comment by lshep** [comment]
I think I have the tryCatch implemented to automate if there was an issue accessing a specific resources' check for updates that it will utilize the cached version -  I want to do testing over the weekend and hope to have it pushed up sometime on monday

**Comment by lshep** [comment]
I pushed up a commit. It won't make it in for tomorrow's report but should be on tue. It wraps the bfcneedsupdate in a trycatch to use the cached version already downloaded if for some reason it can't be checked.

**Comment by LTLA** [comment]
Thanks @lshep. I think that is the best we can do here. The underlying connectivity problem seems to be outside our control, especially if the cause is what I think it is.

**Comment by hypaik** [comment]
Dear developer.

I tried another lab member's computer, and he can load the resource without any problem. 
This morning, I tried other methods, then I can load the resource.
Here, is the method I've used.

1) Delete celldex, SingleR, and reinstall everything.
2) Destroy the directory of Experiment Hub Caches.
3) Deleted *.RData, *.Rhistory files. And use a different working directory.
4) Switch off the ethernet cable and activate the wifi connection of my computer. (My institute blocked AWS for security.)
5) celldex worked properly.

So, I think this issue associated with the firewall of my institute. 
However, it's still unclear how another lab member can load the resource via ethernet cable. 
I think I can figure it out with the IT administrator of my institute.

Now I have a quick question. Do I need to save the downloaded resource as an RData? 
Because the Wifi setting is slow in my institute, it seems uncomfortable to use wifi for every analysis.
